### DISCUSSION TITLE
Can't import worker

### DISCUSSION THREAD

**Discussion by paulrevival** [ujjwalguptaofficial/JsStore#243] [Q&A] [answered]
Hello! Can you help me a bit? :)
I develop Electron app using react-create-app template (without react-rewire & etc..) and I can't require jsstore.worker.js properly :\

I've installed "file-loader" package but it throws an error:
`Cannot find module 'file-loader?name=scripts/[name].[hash].js!jsstore/dist/jsstore.worker.js'`

My code is:

```
const { Connection } = require('jsstore');
const { someTableSchema } = require('../../store/dbTables');

const getWorkerPath = () => {
  return require('file-loader?name=scripts/[name].[hash].js!jsstore/dist/jsstore.worker.js');
};
const workerPath = getWorkerPath().default;
const connection = new Connection(new Worker(workerPath));

const DB = {
  name: 'my-db',
  tables: [
    someTableSchema
  ]
}

async function initDBWorker() {
  try {
    await connection.initDb(DB);
  } catch (ex) {
    console.error(ex);
  }
}

module.exports = {
  initDBWorker,
  DBConnector: connection
};

```
```
const { initDBWorker } = require('./dbWorker');

app.whenReady().then(() => {
  (async () => {
   ... some code
  await initDBWorker();
  })();
});
```

Are there any mistakes? :(

And two more questions (sorry, I'm rather far from DB logics yet):
1. Would jsstore save created db as db-file in `userData` directory? Or how to get the db created on previous app launch?
2. Would jsstore require worker properly after building the app ?? Imho It looks a bit weird way of importing scripts, isn't it? :)

Thank you lots!

**Comment by ujjwalguptaofficial** [comment, marked as answer]
Check if file-loader is properly installed by checking dependencies in `package.json`

You can also run it without worker by using this logic - https://jsstore.net/tutorial/connection/#webpack-without-web-worker


  1. Would jsstore save created db as db-file in userData directory? Or how to get the db created on previous app launch?
  ans - the db is made available to same user when app is launched. You only need to initiate it. 
  2.  Would jsstore require worker properly after building the app ?? Imho It looks a bit weird way of importing scripts, isn't it? :)
  ans - this will be bundled along with your other code. So it will work properly. 
  
 I hope this answer helps you. In doubt - please have  a look at our examples repo - https://github.com/ujjwalguptaofficial/jsstore-examples

**Reply by paulrevival** [reply]
Thx for reply :)

Here's my `package.json` devDeps section and `file-loader` is here (`jsstore` is in dependencies section):
```
"devDependencies": {
    "electron": "^8.2.0",
    "electron-builder": "22.4.1",
    "electron-rebuild": "2.0.1",
    "file-loader": "^6.2.0",
    "husky": "^3.1.0",
    "lint-staged": "^9.5.0",
    "prettier": "^1.19.1",
    "typescript": "^3.8.3"
  }
```

Even if I try to use jsstore without worker, it throws me another error :(
```
ReferenceError: self is not defined
    at Connection.get ([appName]/node_modules/jsstore/dist/jsstore.commonjs2.js:399:36)
    at Connection.module.exports.2.ConnectionHelper.initQueryManager_ ([appName]/node_modules/jsstore/dist/jsstore.commonjs2.js:405:30)
    at ConnectionHelper ([appName]/node_modules/jsstore/dist/jsstore.commonjs2.js:394:18)
    at new Connection ([appName]/node_modules/jsstore/dist/jsstore.commonjs2.js:623:23)
    at Object.<anonymous> ([appName]/src/workers/dbWorker/dbWorker.js:16:20)
```

My code is:

```
const { Connection } = require('jsstore');
const workerInjector = require('sstore/dist/worker_injector');

const connection = new Connection();
connection.addPlugin(workerInjector);
```

I can't use `import` because I want to use `jsstore` in electron main process where I have only commonjs modules with `require` - may it be a problem ?

**Reply by ujjwalguptaofficial** [reply]
You can't use jsstore inside main process. You will have to use inside child process, jsstore is browser technology. Have a look at elctron example - https://github.com/ujjwalguptaofficial/jsstore-examples/tree/master/electron

**Reply by paulrevival** [reply]
Sounds sad :( My app is multi-windowed and I'd like to create a background DB worker that works in main process and is accessible in all child windows of my app without re-connecting in each window where the worker is needed..

**Reply by ujjwalguptaofficial** [reply]
You can create the db in one child process and it will be accessible in all child process since all share single origin.
